gcs/constraints/circuit/circuit.cc

```diff
@@ -0,0 +1,288 @@
+    }
+    else {
+        proof.emit_proof_comment("Contradicting sub-cycle");
+    }
+
+    proof.emit_proof_line(proof_step.str(), ProofLevel::Current);
+}
+
+auto gcs::prevent_small_cycles(
+    const vector<IntegerVariableID> & succ,
+    const PosVarDataMap & pos_var_data,
+    const ConstraintStateHandle & unassigned_handle,
+    State & state) -> Inference
+{
+    auto result = Inference::NoChange;
+    auto & unassigned = any_cast<list<IntegerVariableID> &>(state.get_constraint_state(unassigned_handle));
+    auto n = succ.size();
+    auto end = vector<long>(n, -1);
+    auto known_ends = vector<long>{};
+    auto chain_lengths = vector<long>{};
+
+    for (auto var : unassigned) {
+        state.for_each_value_while(var, [&](Integer val) -> bool {
+            auto j0 = val.raw_value;
+            auto length = 0;
+            if (state.has_single_value(succ[j0]) && (end[j0] < 0)) {
+                auto j = j0;
+                do {
+                    j = state.optional_single_value(succ[j])->raw_value;
+                    length++;
+                    // Need to check this in case alldiff hasn't been fully propagated.
+                    if (j == j0) {
+                        state.infer_true(JustifyExplicitly{[&](Proof & proof) {
+                            output_cycle_to_proof(succ, j0, length, pos_var_data, state, proof);
+                        }});
+                        increase_inference_to(result, Inference::Contradiction);
+                        return false;
+                    }
+                } while (state.has_single_value(succ[j]));
+                end[j0] = j;
+                known_ends.emplace_back(j0);
+                chain_lengths.emplace_back(length);
+            }
+
+            return true;
+        });
+
+        if (result == Inference::Contradiction)
+            return result;
+    }
+
+    while (! known_ends.empty()) {
+        auto i = known_ends.back();
+        known_ends.pop_back();
+        auto length = chain_lengths.back();
+        chain_lengths.pop_back();
+        if (cmp_less(length, succ.size() - 1)) {
+            increase_inference_to(result,
+                state.infer(succ[end[i]] != Integer{i}, JustifyExplicitly{[&](Proof & proof) {
+                    output_cycle_to_proof(succ, i, length, pos_var_data, state, proof, Integer{end[i]}, Integer{i});
+                }}));
+        }
+        else {
+            increase_inference_to(result, state.infer(succ[end[i]] == Integer{i}, JustifyUsingRUP{}));
+        }
+    }
+    return result;
+}
+
+auto gcs::propagate_non_gac_alldifferent(const ConstraintStateHandle & unassigned_handle, State & state) -> innards::Inference
+{
+    auto & unassigned = any_cast<list<IntegerVariableID> &>(state.get_constraint_state(unassigned_handle));
+
+    auto result = Inference::NoChange;
+    vector<pair<IntegerVariableID, Integer>> to_propagate;
+    {
+        // Collect any newly assigned values
+        auto i = unassigned.begin();
```

Comment:
I'd write this as a for loop with no increment condition inside the `for(;;)`. Possibly more readable, or at least idiomatic?